### ISSUE TITLE
Flatpak

### ISSUE THREAD

**Issue by rugk** [wereturtle/ghostwriter#345]
It would be nice if we could get it from [flathub](https://flathub.org/home), i.e. add some flatpak support, so it is easily installable on other distro's than Ubuntu.

**Comment by sbrl** [comment]
Warning: In my experience, flatpak is very unstable - and needlessly sandboxes applications so that they can't write to any files outside of your Documents folder. This should be carefully considered when packaging.

**Comment by rugk** [comment]
Cannot say anything about stability – except that I use some applications fine and it is very obvious it is still in development –, but the sandboxing is a realy nice security feature and not at all useless. But this is the wrong place to argue about such things, so let's keep this issue on topic.

**Comment by wereturtle** [comment]
I actually have been experimenting with Flatpak.  Even though CuteMarked is in Flathub and also uses qtwebkit, I could not even get it or ghostwriter to build for Flatpak with qtwebkit.  And, yes, I was using CuteMarked's JSON manifest straight from the Github repository for Flatpak.  Ugh.

So then I was forced to port from qtwebkit to qtwebengine just to get a Flatpak build.  At last, I fired up my Flatpak build.  Everything looked good, until I loaded a large document (about 150,000+ words).  Chromium (qtwebengine) took *minutes* to load a preview.  It only takes seconds when running a direct install on my machine.

I suspect there is something wrong with the bundled nvidia driver with Flatpak.  Chromium requires OpenGL to render anything, otherwise you will be introduced to a delightful black screen.  I am using the latest version of Flatpak from the official PPA.  Flatpak still has a bit further to go before it's viable, at least for ghostwriter.  I'd be curious to see if people using Intel integrated graphics or AMD experience the same.  I'll be filing a bug report soon at Flatpak, and hopefully the Flatpak gurus can use ghostwriter to test performance better.

On a brighter note, I've used the porting opportunity to play with React and Chromium/qtwebengine together.  Very nice performance boost for doing a live update with super large documents.  I'm surprised various Electron Markdown editors don't already do this, since React is already baked in.  Or maybe it's harder than that?

**Comment by Alexander-Wilms** [comment]
May I ask what error you encountered? Die you have the BaseApp "io.qt.qtwebkit.BaseApp" installed?

**Comment by rugk** [comment]
As for the performance, maybe you encountered https://github.com/flatpak/flatpak/issues/1768. Except of that, I cannot confirm the issue – it's quite fast. (but i don't use nvidia drivers)

Also this issue is not a duplicate.

**Comment by wereturtle** [comment]
> As for the performance, maybe you encountered flatpak/flatpak#1768. Except of that, I cannot confirm the issue – it's quite fast. (but i don't use nvidia drivers)

In my case, if I don't have the live preview open, everything opens quickly.  Even with the live preview open, everything but the live preview will load quickly.  And it's only slow with large files (150,000+ words). Are there any Flatpak apps you've used with Chromium/qtwebengine that have slowed down with large web pages?  I'm curious to know.

> Also this issue is not a duplicate.

Sorry, I should have linked issue #293 when I applied the "duplicate" tag.

**Comment by wereturtle** [comment]
@rugk 

> May I ask what error you encountered? Did you have the BaseApp "io.qt.qtwebkit.BaseApp" installed?

Yes, I did have that installed.  Here are all the steps I did.  I started with copying from CuteMarkEd's manifest.  When that didn't work, I went back to the original CuteMarkEd manifest, and tried to get that one to build first.

As a prerequisite for building CuteMarkEd (as well as ghostwriter), I've installed the qtwebkit base app and the Qt and KF5 Runtimes with the following commands using the latest Flatpak from the
official Ubuntu PPA (version 0.11.7 as of a few weeks ago, but then 0.11.8 came out and that didn't work either when I tried it):

```
$ flatpak remote-add --if-not-exists flathub
https://flathub.org/repo/flathub.flatpakrepo
$ flatpak install flathub org.kde.Platform//5.9
$ flatpak install flathub org.kde.Sdk//5.9
$ flatpak install flathub io.qt.qtwebkit.BaseApp
```

I've then created my build area directory, and beneath that created
app and repo directories.

I am using the io.github.cloose.CuteMarkEd.json and
io.github.cloose.CuteMarkEd.appdata.xml files from
[here](https://github.com/flathub/io.github.cloose.CuteMarkEd).  My directory
structure looks like this:

```
build_dir
-> app/
-> repo/
-> io.github.cloose.CuteMarkEd.json
-> io.github.cloose.CuteMarkEd.appdata.xml
```

Finally, I've tried to build with the following command from within build_dir:

```
$ flatpak-builder --ccache --repo=repo --subject="Build of CuteMarkEd `date`" app io.github.cloose.CuteMarkEd.json --force-clean
```

The first error I get is the following:

```
ln: failed to create symbolic link
'/app/include/QtWebKitWidgets/QWebFrame.h': No such file or directory
```

It appears that in my app/ directory, there is a files/lib directory,
but no files/include directory.  Likewise for
/var/lib/flatpak/app/io.qt.qtwebkit.BaseApp/current/active/files/.
There is only a lib directory with the qtwebkit libraries, but there
is no include directory with the header files.

If I remove the line from io.github.cloose.CuteMarkEd.json in
build-commands that creates the softlink to QWebFrame.h, and allow
qmake to proceed, I end up with the following error:

```
Project ERROR: Unknown module(s) in QT: webkitwidgets
```

I'm wondering if something isn't quite right with qtwebkit repo it's pulling from to build.  I ran into a similar issue when I tried making a manifest for ghostwriter that built qtwebkit first from the same branch.  qtwebkit built fine, but when flatpak-builder went to build ghostwriter, qmake could not find the webkitwidgets module.  The include directory was likewise missing.

The real mystery (at least to me) is how CuteMarkEd was ever successfully built for Flathub in the first place.  I am wondering what installation step I am missing.

If there are any ideas you can offer, I'd appreciate it.  I did try hanging out in the Flatpak IRC channel, but I didn't get a response.  Likewise when I e-mailed the Flatpak mailing list (but I got a "Your message to Flatpak awaits moderator approval" email in return).

Thanks in advance!

**Comment by Alexander-Wilms** [comment]
Hm, I can't reproduce this, I just built it. Does `flatpak list` show `io.qt.qtwebkit.BaseApp/x86_64/stable`? Does uninstalling and reinstalling the BaseApp help?
